### ISSUE TITLE
Add Enum for tuples

### ISSUE THREAD

**Issue by norcalli** [xfix/enum-map#13]
It seems logical that `(A, B) where A: Enum, B: Enum` would be basically the same as a `struct`'s derivation, which is a cartesian product. I don't think we're allowed to implement this ourselves as tuples are always a foreign type, but it could be implemented for up to some length of tuples in the crate.

I personally want this so that I can produce enums via this utility function I've made:

```
#[inline(always)]
pub fn enum_iter<E: enum_map::Enum>() -> impl DoubleEndedIterator<Item = E> + Clone {
    (0..E::LENGTH).map(|i| E::from_usize(i))
}
```

without having to nest the `enum_iter` calls if there's multiple which I want to produce from.

(Side note: the `enum_iter` function might also be useful to have put in.)

I could make an MR if it's an amenable feature to include.

**Comment by xfix** [comment]
I don't think this is possible to implement without `generic_const_exprs` nightly feature if I want to support enum maps of arbitrary types, and I have to no idea whether it would make sense to implement `Enum` only.

That said, a valid workaround if you know what enum cross-product you want would be define your own structure that holds both enums.

```rust
#[derive(Enum)]
struct Product(A, B);
```

Alternatively it's possible to nest enum maps: `EnumMap<A, EnumMap<B, _>>`.
